File: hard/3_three_number_sum/three_number_sum.js
const threeNumberSum = (array, target) => {
	array.sort((a, b) => a - b);
	const triplets = [];
	for (let i = 0; i < array.length - 2; i++) {
		let left = i + 1;
		let right = array.length - 1;
		while (left < right) {
			const currentSum = array[i] + array[left] + array[right];
			if (currentSum === target) {
				triplets.push([array[i], array[left], array[right]]);
				left += 1;
				right -= 1;
			} else if (currentSum < target) {
				left += 1;
			} else if (currentSum > target) {
				right -= 1;
			}
		}
	}
	return triplets;
};

const array = [12, 3, 1, 2, -6, 5, -8, 6];
const target = 0;
console.log(threeNumberSum(array, target));

module.exports = threeNumberSum;

File: hard/3_three_number_sum/three_number_sum.py
def three_number_sum(array, target):
    array.sort()
    triplets = []
    for i in range(len(array)-2):
        left, right = i+1, len(array)-1
        while left < right:
            current_sum = array[i] + array[left] + array[right]
            if current_sum == target:
                triplets.append([array[i], array[left], array[right]])
                left += 1
                right -= 1
            elif current_sum < target:
                left += 1
            elif current_sum > target:
                right -= 1
    return triplets

if __name__ == "__main__":
    array = [12, 3, 1, 2, -6, 5, -8, 6]
    target = 0
    print(three_number_sum(array, target))